### ISSUE TITLE
Not all URLs are found depending on HTML structure

### ISSUE THREAD

**Issue by MichaIng** [lycheeverse/lychee#83]
It seems that parsing a HTML file for links aborts unexpectedly at some point, depending on the format/structure of that HTML file:
```
./lychee -v https://dietpi.com/index.html
./lychee -v https://dietpi.com/docs/index.html
# each
📝 Summary
-------------------
🔍 Total: 0
✅ Successful: 0
⏳ Timeouts: 0
🔀 Redirected: 0
👻 Excluded: 0
🚫 Errors: 0
```
I added some random newlines to the document which lead to two URLs being found.

The first HTML is in non-minified (otherwise exact match) form available here, where at least 6 URLs are found and checked:
```
./lychee -v https://raw.githubusercontent.com/MichaIng/DietPi-Website/master/index.html
📝 Summary
-------------------
🔍 Total: 6
✅ Successful: 6
⏳ Timeouts: 0
🔀 Redirected: 0
👻 Excluded: 0
🚫 Errors: 0
```
But when you check the HTML, there are a lot more links of different kinds. There seem to be something, probably too long lines (?), that makes lychee stop parsing for further URLs, or at least does not find any further, even that they are there.

**Comment by pawroman** [comment]
Thanks for reporting this. It seems that the HTML link extractor has some issues parsing the HTML structure properly. I looked at  https://raw.githubusercontent.com/MichaIng/DietPi-Website/master/index.html.

We currently use an XML parser ([quick-xml](https://github.com/tafia/quick-xml)), and it seems to have issues parsing things like:

```
<script defer src="js/jquery.min.js"></script>
```

Erroring out with 

```
Err(NoEqAfterName(13))
```

and following that, it gives up on parsing prematurely (before reaching the end of the document) with:

```
EndEventMismatch { expected: "link", found: "head" }
```

This is kind of expected with the current implementation, because `<script defer src="js/jquery.min.js"></script>` is not a valid XML construct (the `defer` attribute is missing the value).

The solution to this issue is to switch to a proper HTML parser and avoid parsing HTML as XML (because HTML5 is "kind of XML but not really" and therefore it needs a HTML-aware parser). I think we should give [html5ever](https://github.com/servo/html5ever) a shot -- what do you think, @mre ?

**Comment by pawroman** [comment]
@mre As a temporary quick fix, we can also use: [`reader.check_end_names(false)`](https://docs.rs/quick-xml/0.20.0/quick_xml/struct.Reader.html#method.check_end_names)

This prevents it from erroring if XML is malformed and increases the number of links found in https://raw.githubusercontent.com/MichaIng/DietPi-Website/master/index.html from 6 to 22, parsing the file until the end.

But despite being able to address the issue with a quick fix, my suggestion to switch to a proper HTML parser is still current for the long term.

Let me draft a MR for this.

**Comment by MichaIng** [comment]
Good find. Yes in HTML closing tags or attribute values are not always required, so it makes sense to not try to parse correct XML. Could it make sense to concentrate on `<a href` `<img src` `<link href` `<script src`, i.e. known HTML tag/attribute combinations that are links or asset sources, instead of trying to parse (and by this somehow validate) the whole code? Often HTML snippets are stored as well and combined as part of build/deploy steps, and I'd still want the CSS URL `<link rel="stylesheet" type="text/css" href="https://dietpi.com/css/style.min.css">` being checked even if it is standalone or wrapped into incomplete/invalid HTML code. _Local URL checking (without scheme + domain/hostname) is an own topic._

**Comment by pawroman** [comment]
> Could it make sense to concentrate on e.g. `<a href` `<img src` `<link href` `<script src`, i.e. known HTML tag/attribute combinations that are links or asset sources, instead of trying to parse (and by this somehow validate) the whole code?

It's essentially what we're doing already (for the most part), see: https://github.com/lycheeverse/lychee/blob/b7ab4abb0d058362731dd02fc323ff11ff5b6017/src/extract.rs#L67-L104 and the link extractor will extract links from `<link href="...">`.

**Comment by MichaIng** [comment]
But the attributes are gathered by a parser that expects correct code around it to succeed? What I meant is really a manual e.g. regex text parsing, like `<a[[:blank:]][^>]*href=["']?([^"'[:blank:]])` (cobbled together example, for sure there are better ones 😅). On the other hand, I guess there are existing parsers that can be configured to ignore errors, to likely no need to re-invent the wheel.

**Comment by pawroman** [comment]
> On the other hand, I guess there are existing parsers that can be configured to ignore errors, to likely no need to re-invent the wheel.

Precisely. Parsing HTML using regex is a bad idea in general.

The MR fixing the missing links is up: #92 -- see the code, I have added appropriate tests there.

Note we'll likely need to switch the XML parser to a HTML one, but that's more involved and will take some more time to address (this would fix the missing links for `<script defer src="..."></script>`).

**Comment by MichaIng** [comment]
> Parsing HTML using regex is a bad idea in general.

Depends on the use case and needs. At least it serves great flexibility and one does not rely on upstream fixes or workarounds to overcome missing flexibility and such. But yeah, it could become a maintenance monster as well, I guess.

> The MR fixing the missing links is up: #92

Great work, many thanks! I'm not good in reading Rust, but I see the two elements that allow missing closing tags and empty attribute values 👍.

**Comment by mre** [comment]
Thanks @MichaIng for reporting the issue and @pawroman for proposing a fix.
#92 is merged now and I'll tag 0.4.1 after CI has passed. Does this unblock you @MichaIng?

As for html5ever, I love the idea. @pawroman would you want to work on that?

**Comment by MichaIng** [comment]
I just tested, and the raw HTML is now parsed completely: https://raw.githubusercontent.com/MichaIng/DietPi-Website/master/index.html
What still does not work is the minifed build, the first two tests above are still valid (the third is fixed): https://github.com/lycheeverse/lychee/issues/83#issue-771388796

I tried to re-add body + head tags, re-added line breaks, added quotes around attribute values from the top on to the first URL that is parsed successfully in the formatted HTML, but it seems to stop completely due to errors later in the file.

Btw, is there a way to add/print such internal parser errors?

**Comment by mre** [comment]
> I tried to re-add body + head tags, re-added line breaks, added quotes around attribute values from the top on to the first URL that is parsed successfully in the formatted HTML, but it seems to stop completely due to errors later in the file.

I think this can only be fixed properly once we move to html5ever for parsing the HTML. That's a priority.

> Btw, is there a way to add/print such internal parser errors?

Nope. We could add a `--debug` flag for that, though.

**Comment by MichaIng** [comment]
> Nope. We could add a `--debug` flag for that, though.

If it's only about dropping or printing messages that are received from the parser (or other used libraries) anyway, then that would be great.

**Comment by pawroman** [comment]
My debugging was very low tech, I added a `dbg!` call to diagnose the issue, as well as changed `while let Ok(...) = ` to a plain loop with a match in `extract_links_from_html`. Doing that again (see my branch with the debug changes: https://github.com/pawroman/lychee/commit/4ada976867c41ae70a4dcf3fb014db73bc588ae2) for https://dietpi.com/index.html and https://dietpi.com/docs/index.html I have uncovered one more issue with the current parser:

```
[src/extract.rs:77] err = UnexpectedBang
```

This seems to be caused by the leading `<!doctype html>` in both files. When I removed it, it started working. However, interestingly enough, when I renamed the file from `.html` to `.txt` it found a lot more links... Which means the current parser is really hopeless. I think enough is enough -- I would avoid hacking around the bad parser to support more and more edge cases with parsing HTML as XML.

I actually think that this trick might be a good temporary workaround for you, until we get the proper HTML parser working. E.g. download the HTML files and rename them to have names ending with something other than `.html` (e.g. `.txt`) before checking, to force a plain text mode link search, e.g.:

```
# avoid currently buggy (as of lychee 0.4.1) HTML parsing
# which doesn't like `<!doctype html>` and doesn't find all links
# (this forces a plain text link finding mode)
curl https://dietpi.com/index.html -o index.txt
lychee --verbose index.txt
```

> Nope. We could add a --debug flag for that, though.

@mre To do this properly, we'd have to add some kind of logging mechanism, and dump such things as `debug` log. I would not limit the debug mode just to the parser and link extractor, but might be a good place to start.

I think we could also consider a separate mode to "extract links only" and dump all the links that *would* be checked, without any checking. This would be very helpful to diagnose such issues in the future.

> I think this can only be fixed properly once we move to html5ever for parsing the HTML. That's a priority.

@mre Yes and no. I think that we should provide a flag to force "plaintext mode" for link extraction in all inputs, avoiding HTML and MD parsing. This might be low tech, and we should (of course) aim to extract all the links from both formats...But it would be helpful in situations like these, as an escape hatch for when users hit some issues with more involved parsing.

> As for html5ever, I love the idea. @pawroman would you want to work on that?

Absolutely. I'm happy to pick this up.

**Comment by MichaIng** [comment]
> <!doctype html>

Indeed, changing it to `<!DOCTYPE html>` (capital letters) solves the zero URL finds, I'd never though about that. While a plain text mode wouldn't be bad as well, it isn't able to find local links, so I agree that a HTML parser makes sense to allow handling those. Probably even a combination would be awesome, so that effective links/references can be checked and additionally (optionally) URLs that appear as text or even in comments. But that's a different topic 😉.

**Comment by mre** [comment]
> To do this properly, we'd have to add some kind of logging mechanism, and dump such things as debug log. I would not limit the debug mode just to the parser and link extractor, but might be a good place to start.

Agreed. We could add it as part of integrating html5ever or do it as a separate task. However you prefer.

> I think we could also consider a separate mode to "extract links only" and dump all the links that would be checked, without any checking. This would be very helpful to diagnose such issues in the future.

Good idea. We could add a `--dump` flag for that maybe? Alternative names could be `--diagnose` or `--dry-run`.

> I think that we should provide a flag to force "plaintext mode" for link extraction in all inputs, avoiding HTML and MD parsing.

No objections against adding e.g. a `--format` flag that enforces a certain format. Perhaps my only concern is that adding too many flags could confuse some first-time users. The `--help` could get slightly overwhelming. In light of that, I'd opt for doing html5ever first and keeping that option as a joker for the future. Sounds good?

**Comment by pawroman** [comment]
> Good idea. We could add a `--dump` flag for that maybe? Alternative names could be `--diagnose` or `--dry-run`.

Let's move this discussion to a dedicated PR: #99 

> No objections against adding e.g. a `--format` flag that enforces a certain format. Perhaps my only concern is that adding too many flags could confuse some first-time users. The `--help` could get slightly overwhelming. In light of that, I'd opt for doing html5ever first and keeping that option as a joker for the future. Sounds good?

Sure, sounds very reasonable.

**Comment by pawroman** [comment]
@MichaIng we have added the new HTML parser (available with release `0.5.0`). Would you like to give that a try see if it helped your use cases?

**Comment by MichaIng** [comment]
Awesome, many thanks for the quick implementation, it works perfectly fine now, all (full) URLs are successfully detected. I'll hence mark this as closed.
